decisions/0013-zero-effort-typesafety.md

```diff
@@ -0,0 +1,185 @@
+}
+```
+
+The language service will be implemented to make minimal changes to the source file when auto-annotating route modules.
+It will also translate any references to line/column numbers in the auto-typed code back to the corresponding line/column numbers in the original code
+so that all typechecking and LSP features report warnings and errors relative to the source file.
+
+Unlike a TS plugin, a language service is easily testable in isolation via the methods in the LSP spec.
+
+### `typecheck` command
+
+If you run `tsc` to typecheck your code, `tsc` won't know about our custom language service and will complain that route export args are typed as `any` due to missing type annotations.
+Instead of invoking `tsc` directly as a CLI, React Router will provide a `typecheck` command that invokes the TypeScript typechecker progammatically with our custom language service.
+
+It's worth noting that `tsc` is the TypeScript _compiler_, not just the typechecker.
+Remix and React Router already use Vite to compile and bundle your code.
+So delegating typechecking to TypeScript programmatically is less of a departure from `tsc` than using Vite is.
+
+One limitation of the `typecheck` command is that it assumes that typechecking will be run for the app _independently_ of any other packages or apps in a monorepo.
+Consequently, this approach does not support monorepos with a shared top-level `tsconfig.json` at the root of the monorepo that uses project references to typecheck across packages in the monorepo.
+
+```txt
+root/
+  tsconfig.json 👈 if this uses `references`, that's a problem
+  packages/
+    mylib1/
+    mylib2/
+  apps/
+    myapp1/
+    myapp2/
+```
+
+Instead, each package of a monorepo should have its own `tsconfig.json` and you should avoid a top-level `tsconfig.json` in monorepos.
+If you were relying on project references to get typechecking without needing to run builds for local dependencies, you can instead use [live types via custom `exports` conditions](https://colinhacks.com/essays/live-types-typescript-monorepo).
+
+Note that this only applies to top-level `tsconfig`s in monorepo roots.
+You can freely use project references _within_ the app root to split up the config for different environments like in the [Vite + React + TS template](https://vite.new/react-ts) on Stackblitz.
+
+### Fallback to manual type annotations
+
+While we think this makes the default DX much better for most projects, we need a graceful fallback to provide typesafety for edge cases.
+
+This feature builds on top of the route-specific typegen described in [#0012](./0012-type-inference.md), so
+those typegen files will still be there if you prefer to manually annotate your route exports.
+
+One reason you might want to do this is if your project uses tools that independently inspect types in your app.
+For example, [typescript-eslint](https://typescript-eslint.io/) will think that route export args are typed as `any`.
```

Comment:
I don't know the implications of this exactly, but my understanding is this is a very popular package, so I'm a bit nervous about this disclaimer. Not sure if there are patterns or an eslint plugin or something we can provide to help smooth adoption, but as it's written this basically sounds likes

> if you want to use `typescript-eslint` you don't get automatic typing at all, you have to manually type everything

Is that correct?